# Assignment: Data Wrangling
### `! git clone https://github.com/ds4e/wrangling`
### Do Q1 and Q2, and either Q3 or Q4, for a total of 3 questions.

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

**Question 1 Code:**

In [30]:
import pandas as pd

#1.1 code
df = pd.read_csv("./data/airbnb_hw.csv")

df['Price'] = df['Price'].astype(str).str.replace('$', '').str.replace(',', '')
df['Price'] = df['Price'].str.strip()
df['Price'] = pd.to_numeric(df['Price'], errors='coerce')
missing_values = df['Price'].isna().sum()
print(f"Number of missing values in the Price column: {missing_values}")


#1.2 code:
df = pd.read_csv('./data/mn_police_use_of_force.csv')

df['subject_injury'] = df['subject_injury'].astype('category')
missing_proportion = df['subject_injury'].isna().mean()
cross_tab = pd.crosstab(df['subject_injury'], df['force_type'])

print(f"Proportion of missing values in Subject Injury: {missing_proportion}")
print("\nCross-tabulation of 'subject_injury' and 'force_type':")
print(cross_tab)



Number of missing values in the Price column: 0
Proportion of missing values in Subject Injury: 0.7619342359767892

Cross-tabulation of 'subject_injury' and 'force_type':
force_type      Baton  Bodily Force  Chemical Irritant  Firearm  \
subject_injury                                                    
No                  0          1093                131        2   
Yes                 2          1286                 41        0   

force_type      Gun Point Display  Improvised Weapon  Less Lethal Projectile  \
subject_injury                                                                 
No                             33                 34                       1   
Yes                            44                 40                       2   

force_type      Police K9 Bite  Taser  
subject_injury                         
No                           2    150  
Yes                         44    172  


response_datetime                      problem is_911_call  \
0   2016/01/01 00:47:36         Assault in Progress          Yes   
1   2016/01/01 02:19:34                       Fight           No   
2   2016/01/01 02:19:34                       Fight           No   
3   2016/01/01 02:28:48                       Fight           No   
4   2016/01/01 02:28:48                       Fight           No   
..                  ...                          ...         ...   
95  2016/01/11 19:02:01         Theft - Report Only           No   
96  2016/01/11 23:39:46  Robbery of Biz In Progress          Yes   
97  2016/01/11 23:39:46  Robbery of Biz In Progress          Yes   
98  2016/01/11 23:39:46  Robbery of Biz In Progress          Yes   
99  2016/01/11 23:39:46  Robbery of Biz In Progress          Yes   

   primary_offense subject_injury         force_type   force_type_action  \
0           DASLT1            NaN       Bodily Force  Body Weight to Pin   
1           DISCON            NaN  Chemical Irritant       Personal Mace   
2           DISCON            NaN  Chemical Irritant       Personal Mace   
3           PRIORI            NaN  Chemical Irritant  Crowd Control Mace   
4           PRIORI            NaN  Chemical Irritant  Crowd Control Mace   
..             ...            ...                ...                 ...   
95          SHOPLF            NaN       Bodily Force          Joint Lock   
96          ROBPAG            Yes       Bodily Force  Body Weight to Pin   
97          ROBPAG            Yes       Bodily Force          Joint Lock   
98          ROBPAG            Yes  Improvised Weapon          Flashlight   
99          ROBPAG            Yes       Bodily Force  Body Weight to Pin   

               race     sex   age        type_resistance  precinct  \
0             Black    Male  20.0                 Tensed         1   
1             Black  Female  27.0  Verbal Non-Compliance         1   
2             White  Female  23.0  Verbal Non-Compliance         1   
3             Black    Male  20.0    Commission of Crime         1   
4             Black    Male  20.0    Commission of Crime         1   
..              ...     ...   ...                    ...       ...   
95  Native American  Female  35.0                 Tensed         1   
96            Black    Male  36.0  Verbal Non-Compliance         2   
97            Black    Male  36.0  Verbal Non-Compliance         2   
98            Black    Male  36.0  Verbal Non-Compliance         2   
99            Black    Male  36.0  Verbal Non-Compliance         2   

     neighborhood  
0   Downtown East  
1   Downtown West  
2   Downtown West  
3   Downtown West  
4   Downtown West  
..            ...  
95  Downtown West  
96           Como  
97           Como  
98           Como  
99           Como  

[100 rows x 13 columns]

**Question 1.1:**
I cleaned the Price column by first converting it to a string and removing the dollar sign and commas to ensure all prices were in a standard numeric format. Then, I stripped any extra whitespaces and converted the values to numbers, using errors='coerce' to handle any invalid entries. After cleaning, there were zero missing values, meaning all prices were successfully converted to a numeric format.

**Question 1.2**
I cleaned the subject_injury column by converting it to a categorical type and checking for missing values. I found that 76.2% of the values were missing, which is a big concern because it means most cases do not have clear injury records. The cross-tabulation shows that when force is used, injuries are sometimes recorded, but in many cases, there is no injury reported. However, the high number of missing values raises some questions about whether injuries were not recorded properly, which could indicate gaps in data collection or reporting.

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?
9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

**Q3.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q4.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?